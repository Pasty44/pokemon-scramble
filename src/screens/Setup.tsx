import { useState } from "react";
import pokemon from "pokemon/all";
import { TPokemon } from "pokemon/types";

export interface ISetupProps {
  startGame: (list: Array<TPokemon>) => void;
}

export default function Setup({ startGame }: ISetupProps) {
  const [gens, setGens] = useState<{ [key: string]: boolean }>(() => {
    return Object.keys(pokemon).reduce((acc, curr, index) => {
      if (index === 0) acc[curr] = true;
      else acc[curr] = false;
      return acc;
    }, {} as { [key: string]: boolean });
  });

  const onClickStart = () => {
    if (Object.keys(gens).every((key) => gens[key] === false)) {
      return alert("You need to select at least 1 generation");
    }

    startGame(
      Object.keys(gens).reduce((acc, curr) => {
        if (!!gens[curr]) {
          return [...acc, ...pokemon[curr]];
        } else return acc;
      }, [] as Array<TPokemon>)
    );
  };

  return (
    <>
      <img
        id="logo"
        src={process.env.PUBLIC_URL + "/logo.png"}
        alt="pokemon scramble logo"
      />

      <div id="instructions" className="my-1">
        Select which generations are included in the game. Then try to work out
        the scrambled pokemon name!
      </div>

      <div id="genCheckboxContainer">
        {Object.keys(pokemon).map((key) => (
          <div
            className="genCheckbox noSelect"
            key={key}
            onClick={() => setGens({ ...gens, [key]: !gens[key] })}
            style={{
              border: `3px solid ${!!gens[key] ? "black" : "grey"}`,
              color: !!gens[key] ? "black" : "grey",
            }}
          >
            GEN {key}
          </div>
        ))}
      </div>

      <button id="startBtn" className="my-1 noSelect" onClick={onClickStart}>
        START GAME
      </button>
    </>
  );
}
