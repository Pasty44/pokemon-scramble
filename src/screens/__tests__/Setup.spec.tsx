import renderer from "react-test-renderer";
import Setup from "screens/Setup";

describe("Setup", () => {
  it("should render gen 1 selected only initially", () => {
    const component = renderer.create(<Setup startGame={jest.fn()} />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should alert if no gens are selected", () => {
    jest.spyOn(window, "alert").mockImplementation(() => {});
    const component = renderer.create(<Setup startGame={jest.fn()} />);

    const testInstance = component.root;

    renderer.act(() => {
      testInstance.findByProps({ id: "gen1" }).props.onClick();
    });

    renderer.act(() => {
      testInstance.findByProps({ id: "startBtn" }).props.onClick();
    });

    expect(alert).toBeCalledWith("You need to select at least 1 generation");
  });

  it("should toggle gen div when clicked", () => {
    const component = renderer.create(<Setup startGame={jest.fn()} />);

    const testInstance = component.root;

    expect(
      testInstance
        .findAllByProps({ ["data-id"]: "gen-div" })
        .filter((el) => !el.props.className.includes("disabled")).length
    ).toBe(1);

    renderer.act(() => {
      testInstance.findByProps({ id: "gen2" }).props.onClick();
    });

    expect(
      testInstance
        .findAllByProps({ ["data-id"]: "gen-div" })
        .filter((el) => !el.props.className.includes("disabled")).length
    ).toBe(2);
  });
});
