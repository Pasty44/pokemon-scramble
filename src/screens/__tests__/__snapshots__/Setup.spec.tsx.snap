// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup should render gen 1 selected only initially 1`] = `
Array [
  <img
    alt="pokemon scramble logo"
    id="logo"
    src="/logo.png"
  />,
  <div
    className="my-1"
    id="instructions"
  >
    Select which generations are included in the game. Then try to work out the scrambled pokemon name!
  </div>,
  <div
    id="genCheckboxContainer"
  >
    <div
      className="genCheckbox noSelect"
      data-id="gen-div"
      id="gen1"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid black",
          "color": "black",
        }
      }
    >
      GEN 
      1
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen2"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      2
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen3"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      3
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen4"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      4
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen5"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      5
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen6"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      6
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen7"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      7
    </div>
    <div
      className="genCheckbox noSelect disabled"
      data-id="gen-div"
      id="gen8"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      8
    </div>
  </div>,
  <button
    className="my-1 noSelect"
    id="startBtn"
    onClick={[Function]}
  >
    START GAME
  </button>,
]
`;
